### PR TITLE
MAINT: Cleaner reading and warnings

### PR DESCRIPTION
**Pull request by larsoner**
Inspired by https://mne.discourse.group/t/error-load-fif-file/7218/4:

1. Fixed reading in the case where the tag header is incomplete (e.g., 6 bytes instead of all 16 or 0), on `main` the new test gives the following and passes on this PR (correctly warns):
   ```
   ...
   mne/io/tag.py:168: in _read_tag_header
       return Tag(*struct.unpack(">iIii", s))
   E   struct.error: unpack requires a buffer of 16 bytes
   ```
3. Improved `mne show_fiff` to show sparse matrix dtypes properly:
   ```
   $ mne show_fiff ~/mne_data/MNE-sample-data/subjects/sample/bem/sample-oct-6-src.fif
   ... main:
               3599 = FIFF_MNE_SOURCE_SPACE_DIST (813284b ?1075838980?) =   (187, 180)	0.0047864877¶  (2 ... sparse (csr) shape=(156866, 156866)
   ... PR:
               3599 = FIFF_MNE_SOURCE_SPACE_DIST (813284b >f4)  =   (187, 180)	0.0047864877¶  (2 ... sparse (csr) shape=(156866, 156866)
   ...
   ```
4. Update `constants.py` to have types that are in fiff-constants but are only implicitly in our code
5. Refactored some old cruft `write`/`tag` code to use cleaner conventions and be more DRY

On the problematic file from discourse I get now:
```
/home/larsoner/python/virtualenvs/base/bin/mne:8: RuntimeWarning: FIF tag directory missing at the end of the file (at byte 547940689), possibly corrupted file: EC_sub17_run03.fif
  sys.exit(main())
/home/larsoner/python/virtualenvs/base/bin/mne:8: RuntimeWarning: Invalid tag with only 6/16 bytes at position 547957371
  sys.exit(main())
999  = FIFFB_ROOT
    100  = FIFF_FILE_ID (20b ids)  = {'version': 65538, 'machid': a ... dict len=4
    101  = FIFF_DIR_POINTER (4b >i4)  = [547940689]
    106  = FIFF_FREE_LIST (4b >i4)  = [-1]
    100  = FIFFB_MEAS
        103  = FIFF_BLOCK_ID (20b ids)  = {'version': 65538, 'machid': a ... dict len=4
        101  = FIFFB_MEAS_INFO
            x1047: 0    = ??? (0b nul)
```
which tells us something about the file being corrupted in some way at least rather than just failing with the `unpack` error.

### CODE REVIEW

**Comment by larsoner** [comment]
... and the new warning showed me that we were using `start_file` in `annotations.py` when we should have been `start_and_end_file`, so I've fixed that now, too.

**Review by agramfort — APPROVED** [review]
❤️

**Comment by larsoner** [comment]
Azure failure seems unrelated, feel free to review/merge if you're happy @drammock !

File: mne/io/open.py [hunk 1]
```diff
@@ -162,10 +162,13 @@ def _fiff_open(fname, fid, preload):
     read_slow = True
     if dirpos > 0:
         dir_tag = read_tag(fid, dirpos)
-        if dir_tag is None:
+        if dir_tag is None or dir_tag.data is None:
+            fid.seek(0, 2)
+            sz = fid.tell()
+            extra = "" if sz > dirpos else f" > file size {sz}"
```
Comment (drammock) [inline mne/io/open.py:168]:
```suggestion
            fid.seek(0, 2)  # move to end of file
            size = fid.tell()
            extra = "" if size > dirpos else f" > file size {size}"
```

File: mne/io/tag.py [hunk 2]
```diff
@@ -546,3 +522,28 @@ def _int_item(x):
...
+    matrix_coding = matrix_coding_dict[matrix_coding]
+    matrix_bit_dtype = {
+        FIFF.FIFFT_INT: (4, ">i4"),
+        FIFF.FIFFT_JULIAN: (4, ">i4"),
+        FIFF.FIFFT_FLOAT: (4, ">f4"),
+        FIFF.FIFFT_DOUBLE: (8, ">f8"),
+        FIFF.FIFFT_COMPLEX_FLOAT: (8, ">f4"),
+        FIFF.FIFFT_COMPLEX_DOUBLE: (16, ">f8"),
+    }
+    _check_option("matrix_type", matrix_type, list(matrix_bit_dtype.keys()))
```
Comment (drammock) [inline mne/io/tag.py:547]: [nitpick]
```suggestion
    _check_option("matrix_type", matrix_type, list(matrix_bit_dtype))
```

File: mne/io/tag.py [hunk 3]
```diff
@@ -546,3 +522,28 @@ def _int_item(x):
...
+    matrix_coding = tag.type & 0xFFFF0000
+    if matrix_coding == 0 or tag.size == 0:
+        return None
+    matrix_type = tag.type & 0x0000FFFF
+    matrix_coding_dict = {
+        FIFF.FIFFT_MATRIX: "dense",
+        FIFF.FIFFT_MATRIX | FIFF.FIFFT_SPARSE_CCS_MATRIX: "sparse CCS",
+        FIFF.FIFFT_MATRIX | FIFF.FIFFT_SPARSE_RCS_MATRIX: "sparse RCS",
+    }
+    _check_option("matrix_coding", matrix_coding, list(matrix_coding_dict.keys()))
```
Comment (drammock) [inline mne/io/tag.py:537]: [nitpick, feel free to reject]
```suggestion
    _check_option("matrix_coding", matrix_coding, list(matrix_coding_dict))
```

**Review by drammock — APPROVED** [review]

**Comment by drammock** [comment]
Great improvements. just 3 tiny suggestions for code clarity, all are nitpicky so I'll leave it to you @larsoner if you want to merge as-is or accept (some of) them.

**Pull request merged**